### PR TITLE
[SPARK-46589][CONNECT] Handle a non-existent error class

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to check existence of error class in Scala connect client while converting errors from proto to Spark exceptions. And when an error class doesn't exists, use some legacy error class.

### Why are the changes needed?
To avoid internal errors when the client receives non-existent error class from the connect server, for instance in the case `client 4.0` <-> `server 4.1` as new Spark versions might have new error classes that have been ported to old clients.

### Does this PR introduce _any_ user-facing change?
Yes. After the changes, the connect client (Scala client) won't fail w/ an internal error when it cannot find an error class.

### How was this patch tested?
By existing GAs.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@heyihong @cloud-fan @HyukjinKwon Please, review this PR.

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala [hunk 1]
```diff
@@ -380,9 +388,19 @@ private[client] object GrpcExceptionConverter {
...
    */
-  private def errorParamsToMessageParameters(params: ErrorParams): Map[String, String] =
-    params.errorClass match {
-      case Some(_) => params.messageParameters
-      case None => Map("message" -> params.message)
+  private def errorParamsToMessageParameters(params: ErrorParams): Map[String, String] = {
+    if (isValidErrorClass(params.errorClass)) {
+      params.messageParameters
+    } else {
+      Map("message" -> Option(params.message).getOrElse("No message"))
```
Comment (heyihong) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala:395]: params.message is not null.
```
 Map("message" -> params.message)
```
should be sufficient

**Review by heyihong — COMMENTED** [review]
Early LGTM with one comment

**Review by heyihong — APPROVED** [review]

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala [hunk 2]
```diff
@@ -369,6 +369,14 @@ private[client] object GrpcExceptionConverter {
     errorsToThrowable(0, Seq(builder.build()))
   }
 
+  private def isValidErrorClass(errorClass: Option[String]): Boolean = {
+    errorClass
```
Comment (cloud-fan) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/GrpcExceptionConverter.scala:373]: errorClass.exists(...)

**Comment by MaxGekk** [comment]
> I think the best option is to instantiate the corresponding java exception with error class name, error parameters and the full message.

This require to change behaviour of all sub-classes of `SparkThrowable`. Now we are strict, and require to raise exception w/ correct arguments. You propose to relax all the checks, so, anyone could create exception with any garbage, and we cannot stop that. That's pivot point, and changes design. cc @srielau WDYT?

**Comment by cloud-fan** [comment]
It's more scary to lose the error class name and parameters, which are likely consumed by programs. We should at least use a dedicated java exception to carry the error class name, parameters and the full error message.

**Comment by cloud-fan** [comment]
> You propose to relax all the checks, so, anyone could create exception with any garbage, and we cannot stop that.

I don't think that's an issue. We still require an error class. It's easy to catch malicious code that uses an error class but still pass a dummy error message.

**Comment by MaxGekk** [comment]
> It's more scary to lose the error class name and parameters, which are likely consumed by programs.

This is arguable. I guess user's code probably depends on `SQLSTATE` and wanna see the error message. You cannot get `SQLSTATE` for unknown error class in any case, but the message (it is original message from the server side w/ error class + message + SQLSTATE) can be displayed to users from the legacy error.

**Pull request closed without merging**